flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/source/split/BigQuerySourceSplitAssigner.java

```diff
@@ -0,0 +1,247 @@
+
+    private final ArrayDeque<String> remainingTableStreams;
+    private final List<String> alreadyProcessedTableStreams;
+    private final ArrayDeque<BigQuerySourceSplit> remainingSourceSplits;
+    private final Map<String, BigQuerySourceSplit> assignedSourceSplits;
+    private boolean initialized;
+
+    public BigQuerySourceSplitAssigner(
+            BigQueryReadOptions readOptions, BigQuerySourceEnumState sourceEnumState) {
+        this.readOptions = readOptions;
+        this.remainingTableStreams = new ArrayDeque<>(sourceEnumState.getRemaniningTableStreams());
+        this.alreadyProcessedTableStreams = sourceEnumState.getCompletedTableStreams();
+        this.remainingSourceSplits = new ArrayDeque<>(sourceEnumState.getRemainingSourceSplits());
+        this.assignedSourceSplits = sourceEnumState.getAssignedSourceSplits();
+        this.initialized = sourceEnumState.isInitialized();
+    }
+
+    /**
+     * Reviews the read options argument and see if a query has been configured, in that case run
+     * that query and then return a modified version of the connect options pointing to the
+     * temporary location (project, dataset and table) of the query results.
+     *
+     * @return The BigQuery connect options with the right project, dataset and table given the
+     *     specified configuration.
+     */
+    Optional<BigQueryConnectOptions> fetchOptionsFromQueryRun() {
+        return Optional.ofNullable(this.readOptions.getQuery())
+                // if query is available, execute it using the configured GCP project and gather the
+                // results
+                .flatMap(query -> runQuery(query))
+                // with the query results return the new connection options, fail if the query
+                // failed
+                .map(
+                        result -> {
+                            if (result.getStatus().equals(QueryResultInfo.Status.FAILED)) {
+                                throw new IllegalStateException(
+                                        "The BigQuery query execution failed with errors: "
+                                                + result.getErrorMessages()
+                                                        .orElse(Lists.newArrayList()));
+                            }
+                            String projectId = result.getDestinationProject().get();
+                            String dataset = result.getDestinationDataset().get();
+                            String table = result.getDestinationTable().get();
+                            LOG.info(
+                                    "After BigQuery query execution, switching connect options"
+                                            + " to read from table {}.{}.{}",
+                                    projectId,
+                                    dataset,
+                                    table);
+                            return this.readOptions
```

Comment:
Is this for situations when user passes an entire SQL query and the connector does not use the storage API ? If that is the case then as per my understanding the intermediate data is materialised in GCS bucket but in this case looks like we are reading from a temporary table.